arangod/RocksDBEngine/RocksDBSettingsManager.cpp

```diff
@@ -114,184 +108,185 @@ void RocksDBSettingsManager::retrieveInitialValues() {
-  auto const maxSeqNr = _db->GetLatestSequenceNumber();
-  auto minSeqNr = maxSeqNr;
-  TRI_ASSERT(minSeqNr > 0);
+      LOG_TOPIC("afb17", TRACE, Logger::ENGINES)
+          << "syncing metadata for collection '" << coll->name() << "'";
 
-  rocksdb::TransactionOptions opts;
-  opts.lock_timeout = 50;  // do not wait for locking keys
+      // clear our scratch buffers for this round
+      _scratch.clear();
+      _tmpBuilder.clear();
+      batch.Clear();
 
-  rocksdb::WriteOptions wo;
-  rocksdb::WriteBatch batch;
-  _tmpBuilder.clear();  // recycle our builder
+      auto* rcoll = static_cast<RocksDBCollection*>(coll->getPhysical());
+      rocksdb::SequenceNumber appliedSeq = maxSeqNr;
+      Result res = rcoll->meta().serializeMeta(batch, *coll, force, _tmpBuilder,
+                                               appliedSeq, _scratch);
 
-  auto& dbfeature = _engine.server().getFeature<arangodb::DatabaseFeature>();
-  TRI_ASSERT(!_engine.inRecovery());  // just don't
+      if (res.ok() && batch.Count() > 0) {
+        didWork = true;
 
-  bool didWork = false;
-  auto mappings = _engine.collectionMappings();
+        auto s = _db->Write(wo, &batch);
+        if (!s.ok()) {
+          res.reset(rocksutils::convertStatus(s));
+        }
+      }
 
-  // reserve a bit of scratch space to work with.
-  // note: the scratch buffer is recycled, so we can start
-  // small here. it will grow as needed.
-  _scratch.reserve(128 * 1024);
+      if (res.fail()) {
+        LOG_TOPIC("afa17", WARN, Logger::ENGINES)
+            << "could not sync metadata for collection '" << coll->name()
+            << "'";
+        return res;
+      }
 
-  for (auto const& pair : mappings) {
-    TRI_voc_tick_t dbid = pair.first;
-    DataSourceId cid = pair.second;
-    TRI_vocbase_t* vocbase = dbfeature.useDatabase(dbid);
-    if (!vocbase) {
-      continue;
+      minSeqNr = std::min(minSeqNr, appliedSeq);
     }
-    TRI_ASSERT(!vocbase->isDangling());
-    auto sg = arangodb::scopeGuard([&]() noexcept { vocbase->release(); });
-
-    std::shared_ptr<LogicalCollection> coll;
-    try {
-      coll = vocbase->useCollection(cid, /*checkPermissions*/ false);
-    } catch (...) {
-      // will fail if collection does not exist
-    }
-    // Collections which are marked as isAStub are not allowed to have
-    // physicalCollections. Therefore, we cannot continue serializing in that
-    // case.
-    if (!coll || coll->isAStub()) {
-      continue;
-    }
-    auto sg2 = arangodb::scopeGuard(
-        [&]() noexcept { vocbase->releaseCollection(coll.get()); });
 
-    LOG_TOPIC("afb17", TRACE, Logger::ENGINES)
-        << "syncing metadata for collection '" << coll->name() << "'";
+    if (_scratch.capacity() >= 32 * 1024 * 1024) {
+      // much data in _scratch, let's shrink it to save memory
+      TRI_ASSERT(scratchBufferSize < 32 * 1024 * 1024);
+      _scratch.resize(scratchBufferSize);
+      _scratch.shrink_to_fit();
+      _scratch.clear();
```

Comment:
NIT
`shrink_to_fit` can do nothing

```suggestion
      _scratch = {};
      _scratch.reserve(scratchBufferSize);
```